File: frontend/src/api/aiMedia.ts
import { http, type ApiResponse } from './http'

export type AiMediaKind = 'image' | 'audio'

export interface AiMediaAsset {
  id: number
  media_kind: AiMediaKind
  original_filename: string
  mime_type: string
  byte_size: number
  duration_seconds: number | null
  status: string
  created_time: string
}

export interface AiMediaCapabilities {
  image_enabled: boolean
  audio_enabled: boolean
  max_images: number
  max_image_mb: number
  max_audio_mb: number
  max_audio_seconds: number
  retention_hours: number
  user_quota_mb: number
}

export interface AiMediaUploadOptions {
  courseId?: number | null
  chapterId?: number | null
  durationSeconds?: number | null
  signal?: AbortSignal
  onProgress?: (percent: number) => void
}

export const aiMediaApi = {
  capabilities: () =>
    http.get<ApiResponse<AiMediaCapabilities>>('/ai/media/capabilities'),

  uploadAsset(file: File, options: AiMediaUploadOptions = {}) {
    const formData = new FormData()
    formData.append('file', file, file.name)
    if (options.courseId != null) formData.append('course_id', String(options.courseId))
    if (options.chapterId != null) formData.append('chapter_id', String(options.chapterId))
    if (options.durationSeconds != null) {
      formData.append('duration_seconds', String(options.durationSeconds))
    }

    return http.post<ApiResponse<AiMediaAsset>>('/ai/media/assets', formData, {
      timeout: 120_000,
      signal: options.signal,
      onUploadProgress: (event) => {
        if (!options.onProgress || !event.total) return
        options.onProgress(Math.min(100, Math.round((event.loaded / event.total) * 100)))
      },
    })
  },

  deleteAsset: (assetId: number) =>
    http.delete<ApiResponse<AiMediaAsset>>(`/ai/media/assets/${assetId}`),

  transcribeAsset: (assetId: number, signal?: AbortSignal) =>
    http.post<ApiResponse<{ text: string }>>(
      `/ai/media/assets/${assetId}/transcribe`,
      undefined,
      { timeout: 120_000, signal },
    ),
}

File: backend/app/schemas/ai_media.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict


MediaKind = Literal["image", "audio"]


class AiMediaAssetRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    owner_user_id: int
    course_id: int | None
    chapter_id: int | None
    media_kind: MediaKind
    original_filename: str
    mime_type: str
    byte_size: int
    sha256: str
    duration_seconds: float | None
    status: str
    error_message: str | None
    created_time: datetime
    updated_time: datetime


class AiMediaCapabilities(BaseModel):
    image_enabled: bool
    audio_enabled: bool
    max_images: int
    max_image_mb: int
    max_audio_mb: int
    max_audio_seconds: int
    retention_hours: int
    user_quota_mb: int


class AiMediaTranscription(BaseModel):
    text: str
